src/main/java/org/gingesnap/cdc/consumer/BatchConsumer.java

```diff
@@ -25,15 +27,30 @@ public BatchConsumer(CacheBackend cache, EngineWrapper engine) {
    }
 
    @Override
-   public void handleBatch(List<ChangeEvent<String, String>> batch, DebeziumEngine.RecordCommitter<ChangeEvent<String, String>> recordCommitter) throws InterruptedException {
+   public void handleBatch(List<ChangeEvent<String, String>> batch, DebeziumEngine.RecordCommitter<ChangeEvent<String, String>> recordCommitter) {
       log.info("Processing {} entries", batch.size());
 
       try {
-         for (ChangeEvent<String, String> ev : batch) {
-            // TODO: we may be able to do this in parallel later or asynchronously even
-            CompletionStages.join(process(ev));
-            recordCommitter.markProcessed(ev);
-         }
+         CompletionStage<Void> cs = CompletionStages.allOf(batch.stream()
+                     .reduce(new HashMap<>(), (acc, curr) -> {
+                        acc.compute(curr.key(), (key, prev) -> {
+                           if (prev != null) {
+                              uncheckedCommit(prev, recordCommitter);
+                           }
+                           return curr;
+                        });
+                        return acc;
+                     }, (BinaryOperator<HashMap<String, ChangeEvent<String, String>>>) (first, second) -> {
+                        first.putAll(second);
+                        return first;
+                     }).values().stream()
+                     .map(ev -> process(ev)
+                           .thenApply(ignore -> {
+                              uncheckedCommit(ev, recordCommitter);
+                              return null;
+                           }))
+                     .toArray(CompletionStage[]::new));
+         CompletionStages.join(cs);
```

Comment:
I'm probably missing something, but couldn't we replace the reduce with a traditional loop to improve readability?

```java
         HashMap<String, ChangeEvent<String, String>> acc = new HashMap<>();
         for (ChangeEvent<String, String> curr : batch) {
            acc.compute(curr.key(), (key, prev) -> {
               if (prev != null) {
                  uncheckedCommit(prev, recordCommitter);
               }
               return curr;
            });
         }

         CompletionStages.join(
               CompletionStages.allOf(
                     acc.values().stream()
                           .map(ev -> process(ev)
                                 .thenApply(ignore -> {
                                    uncheckedCommit(ev, recordCommitter);
                                    return null;
                                 }))
                           .toArray(CompletionStage[]::new)
               )
         );
```

Reply:
Definitely much better. I'll apply it.